cmd/snap-confine/snap-confine.apparmor.in

```diff
@@ -303,6 +305,17 @@
     # Allow snap-confine to read snap contexts
     /var/lib/snapd/context/snap.* r,
 
+    # Allow snap-confine to unmount stale mount namespaces.
+    umount /run/snapd/ns/*.mnt,
+    # XXX: It seems the bug is back, we still need this. Without this rule we
+    # get this denial. Note that the code really only (always) unmounts
+    # /run/snapd/ns/$SNAP_NAME.mnt so this is a genuine kernel bug.
+    #
+    # Nov 30 10:26:00 autopkgtest audit[16159]: AVC apparmor="DENIED"
+    # operation="umount" profile="/snap/core/x1/usr/lib/snapd/snap-confine"
+    #
+    umount /,
```

Comment:
I haven't performed a code review for this PR yet, but did look at the approach and discussed with @tyhicks and @jrjohansen (please correct as needed). Here are some important things to note about how snapd is managing namespaces:

* namespaces are not hiearchical and are all single level and live in parallel. snap-confine starts in the default namespace and ultimately uses setns() into the per-snap mount namespace
* using setns() to jump in and out of namespaces (the prior approach) is not how namespaces are meant to be used. setns() is really meant to be one direction
* manipulating mount namespaces can be a very dangerous operation if you can't trust the filesystem inside the mount namespace. Eg, there have been LXC/LXD bugs surrounding this
* manipulating mount namespaces can be tricky if there are processes running inside the mount namespace 
* there are currently no LSM hooks for setns()
* the process manipulating mount namespaces (ie, the setuid snap-confine) is by definition very privileged and so the security policy confining it can only be so strong (eg, all the apparmor mount rules that are currently allowed on '/' (and others), there is no setns() mediation, lack of seccomp filter, ...)


In short, the general approach that this PR (and the subsequent PRs leading up to it) has been discussed with the security team and works within the current limitations of the kernel and keeps the above in mind. Specifically:

* strict confinement snaps running in the namespace are not allowed to use mount or access sensitive files in /proc (eg, nsfs). Importantly, devmode, classic snaps and snaps running on distros without full confinemnet have full access to everything, but they have full root anyway and must be trusted, so no privilege escalation
* nothing is done on the namespace if processes are running within it
* snap-confine has a restrictive AppArmor profile to reduce the attack surface as much as possible. An attacker with total control of snap-confine would be able to exercise significant privilege even under confinement, but the confinement makes attacks and gaining control more difficult 
* snap-confine is defensively coded, undergoes in depth code reviews and is coded to fail closed
* snap-confine doesn't use setns() to jump back and forth within the same process (devmode notwithstanding, but we can look at fixing this separately-- let's not be distracted by that here). Instead it:
    1. starts in the default namespace
    1. forks() a child which does setns() to the snap-specific namespace. The child fork() without exec() is considered sufficient for this specific use case (but exec() is an important security barrier for many other use cases). 
    1. the child interrogates the mount namespace, communicates if changes are needed to the parent and exits
    1. the parent checks if the namespace is in use (via the freezer cgroup). If occupied, does nothing, otherwise discards (tears down) the namespace and reconstructs it

So long as snap-confine continues to setns() in one direction, we severely limit what confined snaps are allowed to do (ie with mount, CLONE_NEWNS, /proc, etc), we correctly detect if processes are running with the mount namespace in a race-free manner, then this is the best we can do.